### PR TITLE
Alternative auto-benchmarking strategy

### PR DESCRIPTION
**Pull request by abadams**
Similar to #2360, but instead of running until a minimum amount of time passes, run until the third best runtime is no more than 3% slower than the best runtime. Potentially faster. 

The buildbots should give us a sense of the performance/flakiness of this approach relative to #2360

### CODE REVIEW

File: tools/halide_benchmark.h [hunk 1]
```diff
@@ -20,19 +59,69 @@ namespace Tools {
...
+BenchmarkResult benchmark(Op op,
+                          const BenchmarkConfig& config = BenchmarkConfig()) {
+    BenchmarkResult result {0, 0, 0};
+
+    uint64_t total_iters = 0;
+    double times[] = {0, 0, 0, 0};
+    for (uint64_t iters_per_sample = config.min_iters;
+         total_iters < config.max_iters;
+         iters_per_sample *= 2) {
+        times[0] = benchmark(1, iters_per_sample, op);
```
Comment (steven-johnson) [inline tools/halide_benchmark.h]: One drawback to this approach is that Really Long Benchmarks now have to run at least 4 times; the GTest-ish approach means that we bail out after a single run.

File: tools/halide_benchmark.h [hunk 2]
```diff
@@ -20,19 +59,69 @@ namespace Tools {
...
-    for (int i = 0; i < samples; i++) {
-        auto t1 = std::chrono::high_resolution_clock::now();
-        for (int j = 0; j < iterations; j++) {
-            op();
+struct BenchmarkConfig {
+    // Terminate when the relative difference between the best runtime
+    // seen and the third-best runtime seen is no more than
+    // this. Controls accuracy. The closer to zero this gets the more
+    // reliable the answer, but the longer it may take to run.
+    double accuracy {0.03};
```
Comment (steven-johnson) [inline tools/halide_benchmark.h]: (Unrelated style nit, but I thought that preferred brace-init syntax elided the space between identifier and brace... at least, that's how I've been doing it.)

File: tools/halide_benchmark.h [hunk 3]
```diff
@@ -20,19 +59,69 @@ namespace Tools {
...
+    
+    // Take at least this much time per sample (seconds). Prevents
+    // benchmarking overhead and clock resolution being a major factor
+    // for very cheap benchmarks.
+    double min_time {0.005};
```
Comment (steven-johnson) [inline tools/halide_benchmark.h]: FWIW, this gives us control flags that have the same names, but different meanings, vs flags in GTest. (One of the goals I was hoping to achieve with this change is to present an interface that would be familiar to users of GTest in terms of the common controllable flags; that may be a bad idea and/or not possible...)

File: apps/resize/resize.cpp [hunk 4]
```diff
@@ -152,7 +152,7 @@ int main(int argc, char **argv) {
         Halide::Runtime::Buffer<>::make_interleaved(in.type(), in.width(), in.height(), in.channels());
     auto out_packed =
         Halide::Runtime::Buffer<>::make_interleaved(out.type(), out.width(), out.height(), out.channels());
-    time = Halide::Tools::benchmark(10, 10, [&]() { resize_fn(in_packed, scale_factor, out_packed); });
+    time = Halide::Tools::benchmark([&]() { resize_fn(in_packed, scale_factor, out_packed); });
```
Comment (steven-johnson) [inline apps/resize/resize.cpp]: Meta-comment: I have a private branch in which  I can run benchmarks with this or GTest-bench easily; I'll compare some problematic cases between this and that (and my original proposal) and see how the results compare.

File: tools/halide_benchmark.h (hunk 1 above)
Comment (abadams) [reply tools/halide_benchmark.h]: Three times usually - it's possible to not enter the loop that adds additional samples if the first three times are similar. This guards against fast tests run during a system hiccup that makes them take unusually long one time. On average this seems to speed things up (~81s in test_performance in this branch on the buildbots, vs ~160s in the other branch, vs ~100s in master)

File: tools/halide_benchmark.h (hunk 3 above)
Comment (abadams) [reply tools/halide_benchmark.h]: Yeah, that's a problem. I was wondering about the interface. Probably should look less similar to gtest.

Comment (steven-johnson) [reply tools/halide_benchmark.h]: ...or more similar :-)

File: tools/halide_benchmark.h [hunk 5]
```diff
@@ -20,19 +59,69 @@ namespace Tools {
...
+
+    uint64_t total_iters = 0;
+    double times[] = {0, 0, 0, 0};
+    for (uint64_t iters_per_sample = config.min_iters;
+         total_iters < config.max_iters;
+         iters_per_sample *= 2) {
+        times[0] = benchmark(1, iters_per_sample, op);
+        times[1] = benchmark(1, iters_per_sample, op);
+        times[2] = benchmark(1, iters_per_sample, op);
+        total_iters += 3;
```
Comment (steven-johnson) [inline tools/halide_benchmark.h]: this should be total_iters += 3 * iters_per_sample

File: tools/halide_benchmark.h [hunk 6]
```diff
@@ -20,19 +59,69 @@ namespace Tools {
...
+        if (times[0] * iters_per_sample < config.min_time) {
+            continue;
         }
-        auto t2 = std::chrono::high_resolution_clock::now();
-        double dt = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count() / 1e6;
-        if (dt < best) best = dt;
+        while (times[0] * (1 + config.accuracy) < times[2] && total_iters < config.max_iters) {
+            times[3] = benchmark(1, iters_per_sample, op);
+            result.samples++;
+            total_iters++;
```
Comment (steven-johnson) [inline tools/halide_benchmark.h]: this should be total_iters += iters_per_sample

File: tools/halide_benchmark.h [hunk 7]
```diff
@@ -20,19 +59,69 @@ namespace Tools {
...
+    // Take at least this much time per sample (seconds). Prevents
+    // benchmarking overhead and clock resolution being a major factor
+    // for very cheap benchmarks.
+    double min_time {0.005};
+
+    // Run at least this many iterations per sample.
+    uint64_t min_iters {1};
+
+    // Run at most this many iterations over all samples.
+    uint64_t max_iters {1000};
```
Comment (steven-johnson) [inline tools/halide_benchmark.h]: This needs to default to a *much* larger value if total_iters really is supposed to track iterations over *all* samples

**Comment by steven-johnson** [comment]
this may be moot since alternative PR has landed?

**Comment by abadams** [comment]
Yep.

**Pull request closed without merging**